### ISSUE TITLE
False inconsistent-return-statements when using unittest.TestCase.fail

### ISSUE THREAD

**Issue by bdrung** [pylint-dev/pylint#8895]
### Bug description

`testcase.py`:

```python
# pylint: disable=missing-docstring
import unittest


class Test(unittest.TestCase):
    def helper(self, success: bool) -> str:
        if success:
            return "success"
        self.fail("It failed")
```

### Configuration

Default configuration.

### Command used

```shell
pylint testcase.py
```

### Pylint output

```shell
************* Module testcase
testcase.py:6:4: R1710: Either all return statements in a function should return an expression, or none of them should. (inconsistent-return-statements)
```

### Expected behavior

No complaint from pylint.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.6
Python 3.11.4 (main, Jun  7 2023, 10:13:09) [GCC 12.3.0]
```

### OS / Environment

Ubuntu 23.10 (mantic)

**Comment by bdrung** [comment]
Do I read #4167 correctly that this bug can be fixed by adding a `NoReturn` type hint to `unittest.TestCase.fail` (and use a pylint version that includes https://github.com/pylint-dev/pylint/pull/4304)?

**Comment by bdrung** [comment]
The typeshed project already declare that `unittest.TestCase.fail` returns `NoReturn`: https://github.com/python/typeshed/blob/0d8a6bc2004b7b6d8335e35b017438125c305264/stdlib/unittest/case.pyi#L253

**Comment by Pierre-Sassoulas** [comment]
pylint do not take the pyi file into account in 2.17.5, this should be available in 3.0 alpha 7 hen we release it though https://github.com/pylint-dev/pylint/issues/4987.
